myhpom/tasks.py

```diff
@@ -0,0 +1,119 @@
+        * line_id = the id of the CloudFactory production line to submit this request to.
+        * document_host = the host (incl. protocol/scheme) to use for the DocumentURL
+        * callback_url = the URL to use with the callback request
+
+        ## Task Process
+        - create an expiring DocumentUrl for the AdvanceDirective
+        - submit the DocumentUrl to the CloudFactory (test) endpoint
+        - store the status of the submitted document
+        - if the AdvanceDirective no longer exists, abort the task without an Exception.
+            - the AdvanceDirective might no longer exist by the time celery picks it up,
+            - or the document it points to might have been removed.
+        - if an error occurs in this process, send support email.
+        """
+        cf_run = CloudFactoryRun.objects.create(line_id=line_id, callback_url=callback_url or "")
+
+        # If the AdvanceDirective has been deleted, abort the task with a message
+        # -- no need to admin support email, this is an expected case.
+        try:
+            ad = AdvanceDirective.objects.get(id=ad_id)
+        except:
+            cf_run.status = 'ABORTED'
+            cf_run.message = 'AdvanceDirective(id=%r) no longer exists.' % ad_id
+            cf_run.save()
+            return cf_run.data
+
+        # Any exception in the rest of the task should result in support email
+        document_url = DocumentUrl.objects.create(advancedirective=ad)
+        CloudFactoryUnit.objects.create(
+            run=cf_run, input=self.create_unit_input(document_url, document_host)
+        )
+        response = self.post_run(cf_run.post_data)
+        response_data = response.json()
+
+        # if the run could not be created (status_code != 201), send support email
+        # -- we need to understand why the the run could not be created at CloudFactory.
+        if response.status_code != 201:
+            cf_run.__dict__.update(
+                status=str(response.status_code), message=response_data.get('message')
+            )
+            cf_run.save()
+            raise ValueError(
+                "%s\n\n== RUN POST DATA ==\n%s\n\n== FULL RUN DATA ==\n%s"
+                % (
+                    response_data['message'],
+                    json.dumps(cf_run.post_data, indent=2),
+                    json.dumps(cf_run.data, indent=2),
+                )
+            )
+
+        cf_run.__dict__.update(
+            run_id=response_data['id'],
+            status=response_data['status'],
+            created_at=response_data['created_at'],
+        )
+        cf_run.save()
+        return cf_run.data  # json-serializable by design
+
+    # separating this so it can be mocked during tests, and not actually hit CloudFactory then.
+    def post_run(self, post_data):
+        return requests.post(settings.CLOUDFACTORY_API_URL + '/runs', json=post_data)
+
+    def create_unit_input(self, document_url, document_host):
```

Comment:
This is the big design question on this PR: Should the CloudFactory models be tied to the needs of the AdvanceDirective, or generalized and decoupled? I designed these models to be generalized and decoupled, even though it's a _little_ more work (not much: I've already created the admin in MH-254, and it's fine). 

So, some push back: It does make sense to me to keep the CloudFactoryRun model de-coupled from the specific requirements of the AdvanceDirectives, and closely tied to the interface provided by CloudFactory. That way, 

(1) if they change their interface, we can more easily change these models and the tasks. The less entanglement we have with our internal models, the easier that will be. 

(2) we can use these models for other CloudFactory production tasks. Not that we envision additional uses of CloudFactory on this project, but who knows? Shane is spending a lot of money on them, and he might think of other ways to use the time he's paying for. 

(3) I'm also thinking in terms of extracting a cloudfactory_api package that can be used on other projects. Not that Caktus envisions working with CloudFactory on other projects, but who knows? If we start doing anything having to do with machine learning, it might end up making sense to employ CloudFactory to train our models (assuming a client will pay for that). (Maybe we can even sell Shane on doing some big data / ML with these Advance Directives....)

There's always the question of whether you should design for the general case or the specific one. We usually stay pretty close to the specific case. But here I think we should keep it generalized & decoupled.

Reply:
speculative, but laudable goals! If we break things out for reused I'd agree with the tighter modeling you've striven for in this PR...I'd probably go for python based generic classes rather than django based so that it'd be useful more generally.